### PR TITLE
rgw: bucket index check in radosgw-admin removes valid index.

### PR DESCRIPTION
**Pull request by zhangsw**
Fixes: http://tracker.ceph.com/issues/18470

Signed-off-by: Zhang Shaowen <zhangshaowen@cmss.chinamobile.com>

### CODE REVIEW

**Comment by zhangsw** [comment]
@cbodley please help review this ,thanks~

**Comment by cbodley** [comment]
looks okay to me - @yehudasa can you double check?

**Comment by theanalyst** [comment]
there is a merge commit in this pr which should be dropped

**Comment by zhangsw** [comment]
@theanalyst I've rebased it and resolved the conflicts

**Comment by yehudasa** [comment]
@zhangsw how did you test it?

**Comment by zhangsw** [comment]
@yehudasa @cbodley I'm sorry that I missed checking whether the bad index removed. I just checked those valid index were not removed. And there is an another bug when removing bad index. I've updated this pr to ensure those bad index will be removed. I've tested it by modifying complete multipart code to leave  the keys those should be removed, and this pr could remove these keys.

**Comment by zhangsw** [comment]
@cbodley @yehudasa ping..

File: src/rgw/rgw_bucket.cc [hunk 1]
```diff
@@ -1028,15 +1030,9 @@ int RGWBucket::check_bad_index_multipart(RGWBucketAdminOpState& op_state,
...
-      rgw_obj obj(bucket, ent.key);
-      obj.key.ns.clear();
-
-      rgw_obj_index_key key;
-      obj.key.get_index_key(&key);
-
-      string oid = key.name;
+      rgw_obj_index_key key = iter->key;
+      rgw_obj obj(bucket, key);
+      string oid = obj.key.name;
```
Comment (cbodley) [inline src/rgw/rgw_bucket.cc]: is there a reason this can't use `rgw_obj::get_oid()`?

Comment (zhangsw) [inline src/rgw/rgw_bucket.cc]: @cbodley Yes, we can also use that. Maybe get_oid() is more readable and clear here. I've changed it

**Comment by cbodley** [comment]
thanks @zhangsw, i was able to reproduce the issue by modifying the multipart complete as you suggested:
```patch
--- a/src/rgw/rgw_op.cc
+++ b/src/rgw/rgw_op.cc
@@ -4947,7 +4947,7 @@ void RGWCompleteMultipart::execute()
       rgw_obj_index_key remove_key;
       src_obj.key.get_index_key(&remove_key);
 
-      remove_objs.push_back(remove_key);
+      //remove_objs.push_back(remove_key);
 
       ofs += obj_part.size;
       accounted_size += obj_part.accounted_size;
```
after a multipart upload to `BUCKET/big.iso`, the `_multipart_big.iso.2~rNBlqXA8BoV1OKRm7z5jY2dviJ_X-g4.XX` objects were still listed in the output of `radosgw-admin bi list --bucket=BUCKET`. the `radosgw-admin bucket check --bucket=BUCKET` command was able to identify all of those objects, and the `--fix` command successfully removed them from the index. a final `radosgw-admin bi list --bucket=BUCKET` command confirmed that `big.iso` was the only object remaining

**Comment by cbodley** [comment]
@yehudasa can you take another look, and merge if you're happy with it?

**Comment by zhangsw** [comment]
@yehudasa ping..

**Comment by yehudasa** [comment]
@zhangsw I will try to take a look at this one this week

**Comment by yehudasa** [comment]
@zhangsw congratulations on being the first person to send a PR to ragweed!

**Comment by yehudasa** [comment]
@zhangsw for some reason I commented on the wrong PR, the comment above referred to the ragweed PR .

**Comment by zhangsw** [comment]
@yehudasa I also wanted to use bi_list，but I didn't find it in rest admin api. Maybe we should add it.

**Pull request merged**